app/classes/textile.rb

```diff
@@ -301,36 +315,59 @@ def strip_out_sp_cfr_and_sensu(str)
       sub(/ sp\.$/, "")
   end
 
-  # Textile should be able to create internal links to Glossary Terms
-  # See https://www.pivotaltracker.com/story/show/76993598
-  # Convert _object name_ and _object id_ in a textile string.
-  def check_other_links!
+  # rubocop:disable Style/RegexpLiteral
+  # cop gives false positive
+  OTHER_LINK_PATTERN = /
+    (?<prefix> ^|\W)
+    (?: _+)
+    (?<type>
+      [a-zA-Z]+ # model name
+      (?: _[a-zA-Z]+)? # optionally including underscores
+    )
+    \s+
+    (?<id> [^_\s](?:[^_\n]+[^_\s])?) # id -- integer or string
+    (?: _+)
+
+    (?! (?: \w|<\/[a-z]+>)) # discard if followed by word char or html close tag
+  /x
+  # rubocop:enable Style/RegexpLiteral
+
+  OTHER_LINK_TYPES = [
+    ["comment"],
+    %w[glossary_term term],
+    %w[image img],
+    %w[location loc],
+    ["name"],
+    %w[observation obs ob],
+    %w[project proj],
+    %w[species_list spl],
+    ["user"]
+  ].freeze
+
+  # Convert _object name_ and _object id_ to a textile string.
+  def convert_other_links_to_tagged_objects!
     gsub!(OTHER_LINK_PATTERN) do |orig|
-      result = orig
-      prefix = Regexp.last_match(1)
-      type = Regexp.last_match(2)
-      id = Regexp.last_match(3)
-      matches = [
-        ["comment"],
-        %w[image img],
-        %w[location loc],
-        ["name"],
-        %w[observation obs ob],
-        %w[project proj],
-        %w[species_list spl],
-        ["user"]
-      ].select { |x| x[0] == type.downcase || x[1] == type.downcase }
-      if matches.length == 1
-        label = (/^\d+$/.match?(id) ? "#{type} #{id}" : id)
-        result =
-          "#{prefix}x{#{matches.first.first.upcase} __#{label}__ }{ #{id} }x"
+      prefix = $LAST_MATCH_INFO[:prefix]
+      type = $LAST_MATCH_INFO[:type]
+      id = $LAST_MATCH_INFO[:id]
+      matches = OTHER_LINK_TYPES.
+                select { |x| x[0] == type.downcase || x[1] == type.downcase }
+
+      if matches.one?
+        "#{prefix}x{#{matches.first.first.upcase} " \
+        "__#{tagged_object_label(type, id)}__ }{ #{id} }x"
+      else
+        orig
```

Comment:
I hate to complain, and I may well have been the one to write it this way originally, but...  This loop looks expensive, and more importantly unnecessary.  How about something like this instead:

```
CLEAN_TYPES = {
  "image": "IMAGE",
  "img": "IMAGE",
  "location": "LOCATION",
  "loc": "LOCATION",
  ...
}.freeze
if clean_type = CLEAN_TYPES[type.downcase]
  label = tagged_object_label(type, id)
  "#{prefix}x{#{clean_type}  __#{label}__}{#{id}}x"
else
  orig
end
```

Does that make sense?

Reply:
Complaints are great! I'll look at this tomorrow. Too tired tonight.

Reply:
@pellaea: Another good idea. Thanks again for the complaint.
Here's what I've done: (pasting the whole thing here so that you get it via email):
```ruby
  MARKED_TYPE_TO_TAGGED_TYPE = {
    comment: "COMMENT",
    glossary_term: "GLOSSARY_TERM",
    image: "IMAGE",
    img: "IMAGE",
    location: "LOCATION",
    loc: "LOCATION",
    name: "NAME",
    term: "GLOSSARY_TERM",
    ob: "OBSERVATION",
    obs: "OBSERVATION",
    observation: "OBSERVATION",
    project: "PROJECT",
    species_list: "SPECIES_LIST",
    spl: "SPECIES_LIST",
    user: "USER"
  }.freeze

  # Convert _object name_ and _object id_ to a textile string.
  def convert_other_links_to_tagged_objects!
    gsub!(OTHER_LINK_PATTERN) do |orig|
      prefix = $LAST_MATCH_INFO[:prefix]
      marked_type = $LAST_MATCH_INFO[:marked_type]
      id = $LAST_MATCH_INFO[:id]

      tagged_type = MARKED_TYPE_TO_TAGGED_TYPE[marked_type.downcase.to_sym]
      next(orig) unless tagged_type

      label = tagged_object_label(marked_type, id)
      "#{prefix}x{#{tagged_type} __#{label}__ }{ #{id} }x"
    end
  end
```

Reply:
Looks great.  (By the way, much better name for the hash than my suggestion. :)